server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java

```diff
@@ -246,59 +250,59 @@ private void loadRepositoryData(SnapshotsInProgress snapshotsInProgress, Snapsho
-                    }
-                    final long startTime = snapshotInfo.startTime();
-                    final long endTime = snapshotInfo.endTime();
-                    assert endTime >= startTime || (endTime == 0L && snapshotInfo.state().completed() == false)
-                        : "Inconsistent timestamps found in SnapshotInfo [" + snapshotInfo + "]";
-                    builder.add(new SnapshotStatus(new Snapshot(repositoryName, snapshotId), state,
-                        Collections.unmodifiableList(shardStatusBuilder), snapshotInfo.includeGlobalState(),
-                        startTime,
-                        // Use current time to calculate overall runtime for in-progress snapshots that have endTime == 0
-                        (endTime == 0 ? threadPool.absoluteTimeInMillis() : endTime) - startTime));
+                if (snapshotsInProgress.snapshot(new Snapshot(repositoryName, snapshotId)) == null) {
+                    snapshotIdsToLoad.add(snapshotId);
                 }
             }
-            listener.onResponse(new SnapshotsStatusResponse(Collections.unmodifiableList(builder)));
-        }, listener::onFailure), threadPool.generic(), null);
-    }
 
-    /**
-     * Retrieves snapshot from repository
-     *
-     * @param snapshotsInProgress snapshots in progress in the cluster state
-     * @param repositoryName      repository name
-     * @param snapshotId          snapshot id
-     * @return snapshot
-     * @throws SnapshotMissingException if snapshot is not found
-     */
-    private SnapshotInfo snapshot(SnapshotsInProgress snapshotsInProgress, String repositoryName, SnapshotId snapshotId) {
-        List<SnapshotsInProgress.Entry> entries =
-            SnapshotsService.currentSnapshots(snapshotsInProgress, repositoryName, Collections.singletonList(snapshotId.getName()));
-        if (entries.isEmpty() == false) {
-            return new SnapshotInfo(entries.iterator().next());
-        }
-        return repositoriesService.repository(repositoryName).getSnapshotInfo(snapshotId);
+            if (snapshotIdsToLoad.isEmpty()) {
+                listener.onResponse(new SnapshotsStatusResponse(Collections.unmodifiableList(builder)));
+            } else {
+                final List<SnapshotStatus> threadSafeBuilder = Collections.synchronizedList(builder);
+                repositoriesService.repository(repositoryName)
+                    .getSnapshotInfo(new GetSnapshotInfoContext(snapshotIdsToLoad, true, () -> false,
+                        (context, snapshotInfo) -> {
+                            List<SnapshotIndexShardStatus> shardStatusBuilder = new ArrayList<>();
+                            final Map<ShardId, IndexShardSnapshotStatus> shardStatuses;
+                            try {
+                                shardStatuses = snapshotShards(repositoryName, repositoryData, snapshotInfo);
```

Comment:
Could/Should make this logic that fetches the shard level meta async in a follow-up as well so it also runs on the snapshot meta pool instead of putting this long running task on it as it does now. I realize that such a slow task may interfere with other snapshot related operations for a bit here but I figured it's not bad to the point that I'd want to fork this off the the `GENERIC` pool again until that follow-up happens (it'll happen right after anyway and flooding the generic pool with these tasks like we do today has its own set of problems also).